compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala

```diff
@@ -136,10 +144,11 @@ object TypeTestsCasts {
             // See TypeComparer#either
             recur(tp1, P) && recur(tp2, P)
           case _ =>
-            // first try withou striping type parameters for performance
+            // always false test warnings are emitted elsewhere
             X.classSymbol.exists && P.classSymbol.exists && !X.classSymbol.asClass.mayHaveCommonChild(P.classSymbol.asClass) ||
-            isClassDetermined(X, tpe)(ctx.fresh.setNewTyperState()) ||
-            isClassDetermined(stripTypeParam(X), tpe)(ctx.fresh.setNewTyperState())
+            // first try without striping type parameters for performance
+            isClassDetermined(X, tpe)(ctx.fresh.setNewTyperState().setFreshGADTBounds) ||
```

Comment:
Instead of `setNewTyperState` you should be able to use `setExploreTyperState`: this is a less expensive version which assumes the typerstate will never be committed.

Reply:
Also, since all calls to `isClassDetermined` need to use a fresh typer state, I recommend moving that logic inside `isClassDetermined` itself.